diff --git a/Makefile.am b/Makefile.am
index e2c1bd8..9b7bbda 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -29,7 +29,8 @@ memcached_SOURCES = memcached.c memcached.h \
                     authfile.c authfile.h \
                     restart.c restart.h \
                     proto_text.c proto_text.h \
-                    proto_bin.c proto_bin.h
+                    proto_bin.c proto_bin.h \
+                    litesys.c litesys.h
 
 if BUILD_SOLARIS_PRIVS
 memcached_SOURCES += solaris_priv.c
diff --git a/configure.ac b/configure.ac
index a0851f2..9863cce 100644
--- a/configure.ac
+++ b/configure.ac
@@ -235,6 +235,10 @@ if test "x$enable_proxy_uring" = "xyes"; then
     CPPFLAGS="-Ivendor/liburing/src/include $CPPFLAGS"
 fi
 
+# Add vendor/LiteSys include directory and linker flags
+CPPFLAGS="-Ivendor/LiteSys $CPPFLAGS"
+LIBS="-Wl,-rpath,vendor/LiteSys/build -Lvendor/LiteSys/build -lembedded_lite_memcached -lstdc++ -lglog $LIBS"
+
 AM_CONDITIONAL([BUILD_DTRACE],[test "$build_dtrace" = "yes"])
 AM_CONDITIONAL([DTRACE_INSTRUMENT_OBJ],[test "$dtrace_instrument_obj" = "yes"])
 AM_CONDITIONAL([ENABLE_SASL],[test "$enable_sasl" = "yes"])
diff --git a/litesys.c b/litesys.c
new file mode 100644
index 0000000..fe06183
--- /dev/null
+++ b/litesys.c
@@ -0,0 +1,57 @@
+#include "litesys.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "memcached.h"
+
+typedef struct lite_item {
+  uint8_t nkey;
+  uint8_t *key;
+  uint8_t flags[2];
+  uint8_t *value;
+  uint32_t nbytes;
+} lite_item;
+
+void LiteFlushWriteBuffer(void *conn) {
+  // TODO: Implement
+}
+
+void LiteReinstallClientEventHandler(void *client, int install) {
+  // TODO
+}
+
+void LiteReinstallListenerEventHandler(void *listener, int install) {
+  // TODO
+}
+
+void LiteQueueEmbeddedRequest(void *conn_info, void *it_, int is_success) {
+  item *it = (item *)it_;
+  lite_item *req = malloc(sizeof(lite_item));
+
+  req->nkey = it->nkey;
+  const char *key = ITEM_key(it);
+  while (!key[req->nkey - 1]) req->nkey--;
+  req->key = malloc(req->nkey);
+  memcpy(req->key, key, req->nkey);
+
+  itoa_u32(*((uint32_t *) ITEM_suffix(it)), (char *)req->flags);
+
+  if ((it->it_flags & ITEM_CHUNKED) != 0) {
+    fprintf(stderr, "Chunked item not supported by embedded mode\n");
+    return;
+  }
+
+  req->nbytes = it->nbytes - 2;
+  req->value = malloc(req->nbytes);
+  memcpy(req->value, ITEM_data(it), req->nbytes);
+
+  LiteProcessRequest(conn_info, req, is_success);
+}
+
+void LiteRequestDestructor(void *request) {
+  lite_item *it = (lite_item *)request;
+  free(it->key);
+  free(it->value);
+  free(it);
+}
diff --git a/litesys.h b/litesys.h
new file mode 100644
index 0000000..7111dcd
--- /dev/null
+++ b/litesys.h
@@ -0,0 +1,13 @@
+#include "embedded_lite.h"
+
+void LiteFlushWriteBuffer(void *conn);
+
+void LiteRequestDestructor(void *request);
+
+// uninstall is 0 if the event handler is being uninstalled, 1 if it is being
+// installed
+void LiteReinstallClientEventHandler(void *client, int install);
+
+void LiteReinstallListenerEventHandler(void *listener, int install);
+
+void LiteQueueEmbeddedRequest(void *conn_info, void *it, int is_success);
\ No newline at end of file
diff --git a/memcached.c b/memcached.c
index 743a75a..0b95c49 100644
--- a/memcached.c
+++ b/memcached.c
@@ -62,6 +62,8 @@
 #include <sys/sysctl.h>
 #endif
 
+#include "litesys.h"
+
 /*
  * forward declarations
  */
@@ -937,6 +939,7 @@ static void conn_close(conn *c) {
         stop_assoc_maintenance_thread();
     }
 #endif
+    LiteUnregisterClient(c->sfd);
     close(c->sfd);
     c->close_reason = 0;
     pthread_mutex_lock(&conn_lock);
     allow_new_conns = true;
@@ -1712,6 +1715,8 @@ enum store_item_type do_store_item(item *it, int comm, conn *c, const uint32_t h
             stored, comm, ITEM_key(it), it->nkey, it->nbytes, it->exptime,
             ITEM_clsid(it), c->sfd);
 
+    LiteQueueEmbeddedRequest(c->lite_data, it, 1);
+
     return stored;
 }
 
@@ -3618,6 +3623,7 @@ static int server_socket(const char *interface,
 #endif
             listen_conn_add->next = listen_conn;
             listen_conn = listen_conn_add;
+            LiteRegisterListener(sfd, listen_conn, 0); // TODO: replay
         }
     }
 
@@ -4203,6 +4209,7 @@ static void remove_pidfile(const char *pid_file) {
 }
 
 static void sig_handler(const int sig) {
+    LiteSignalHandler(sig);
     stop_main_loop = EXIT_NORMALLY;
     printf("Signal handled: %s.\n", strsignal(sig));
 }
@@ -4656,6 +4663,8 @@ static int _mc_meta_load_cb(const char *tag, void *ctx, void *data) {
 }
 
 int main (int argc, char **argv) {
+    LiteInit(argv[0], LiteRequestDestructor, LiteFlushWriteBuffer,
+             LiteReinstallClientEventHandler, LiteReinstallListenerEventHandler);
     int c;
     bool lock_memory = false;
     bool do_daemonize = false;
@@ -6129,6 +6138,8 @@ int main (int argc, char **argv) {
     /* Initialize the uriencode lookup table. */
     uriencode_init();
 
+    LiteFullStartListening();
+
     /* enter the event loop */
     while (!stop_main_loop) {
         if (event_base_loop(main_base, EVLOOP_ONCE) != 0) {
diff --git a/memcached.h b/memcached.h
index 7a577f4..93a1387 100644
--- a/memcached.h
+++ b/memcached.h
@@ -776,6 +776,8 @@ struct _io_pending_t {
  * The structure representing a connection into memcached.
  */
 struct conn {
+    void *lite_data;
+
     sasl_conn_t *sasl_conn;
     int    sfd;
     bool sasl_started;
diff --git a/thread.c b/thread.c
index dc59062..d05c858 100644
--- a/thread.c
+++ b/thread.c
@@ -29,6 +29,8 @@
 #include <openssl/ssl.h>
 #endif
 
+#include "litesys.h"
+
 #define ITEMS_PER_ALLOC 64
 
 /* An item in the connection queue. */
@@ -588,6 +590,7 @@ static void thread_libevent_process(evutil_socket_t fd, short which, void *arg)
                 } else {
                     c->thread = me;
                     conn_io_queue_setup(c);
+                    c->lite_data = LiteRegisterClient(c->sfd, c);
 #ifdef TLS
                     if (settings.ssl_enabled && c->ssl != NULL) {
                         assert(c->thread && c->thread->ssl_wbuf);
diff --git a/vendor/LiteSys/build b/vendor/LiteSys/build
new file mode 120000
index 0000000..ae81360
--- /dev/null
+++ b/vendor/LiteSys/build
@@ -0,0 +1 @@
+../../../lite-version-ascii-embedded/build/
\ No newline at end of file
diff --git a/vendor/LiteSys/embedded_lite.h b/vendor/LiteSys/embedded_lite.h
new file mode 120000
index 0000000..87eae16
--- /dev/null
+++ b/vendor/LiteSys/embedded_lite.h
@@ -0,0 +1 @@
+../../../lite-version-ascii-embedded/Lite/include/embedded_lite.h
\ No newline at end of file
