diff --git a/.gitignore b/.gitignore
index 1b34a3e..0049efd 100644
--- a/.gitignore
+++ b/.gitignore
@@ -8,3 +8,4 @@ tmp.*
 .DS_Store
 data/
 *~
+*.txt
\ No newline at end of file
diff --git a/src/Makefile b/src/Makefile
index be3a690..1d35574 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -49,4 +49,4 @@ distclean: clean
 	cd ../vendor/leveldb; make clean
 
 ../vendor/libleveldb.a:
-	cd ../vendor/leveldb; make && cp out-static/libleveldb.a ..
+	cd ../vendor/leveldb; mkdir -p build && cd build; cmake -DCMAKE_BUILD_TYPE=Release .. && cmake --build .; cp libleveldb.a ../..
diff --git a/src/rl.cpp b/src/rl.cpp
index 72d9685..c0a8fd7 100644
--- a/src/rl.cpp
+++ b/src/rl.cpp
@@ -14,7 +14,6 @@
 #include <sys/socket.h>
 
 #include <algorithm>
-
 #include "rl_util.h"
 #include "rl_server.h"
 #include "rl_connection.h"
@@ -33,6 +32,7 @@ int main(int argc, char** argv) {
 
     int port=8323;
     int db_num=0;
+    size_t write_buffer_size=leveldb::Options().write_buffer_size;
 
     RLRequest::init_cmd_map();
 
@@ -40,7 +40,7 @@ int main(int argc, char** argv) {
     memset(data_dir,0,128);
     strncpy(data_dir,"redis.db",8);
 
-    while ((ch = getopt(argc, argv, "hdr:H:P:D:M:")) != -1) {
+    while ((ch = getopt(argc, argv, "hdr:H:P:D:M:B:")) != -1) {
         switch (ch) {
         case 'h':
             printf("Usage:\n\t./redis-leveldb [options]\n");
@@ -51,6 +51,7 @@ int main(int argc, char** argv) {
             printf("\t-P port:\t listen port\n");
             printf("\t-D data-dir:\t data dir\n");
             printf("\t-M number:\t DB count(run in multi-db mode)\n");
+            printf("\t-B size:\t write buffer size(Bytes)\n");
             exit(0);
         case 'd':
             daemon_flag = 1;
@@ -85,6 +86,18 @@ int main(int argc, char** argv) {
                 exit(1);
             }
             break;
+        case 'B':
+            if(std::find_if(optarg,optarg+strlen(optarg),
+                            std::not1(std::ptr_fun(isdigit)))!=optarg+strlen(optarg)){
+                printf("Bad write buffer size(-B) value(must be a num)\n");
+                exit(1);
+            }
+            write_buffer_size=strtol(optarg,NULL,10);
+            if(write_buffer_size<1){
+                printf("Bad write buffer size(-B) value(must be a num)\n");
+                exit(1);
+            }
+            break;
         case '?':
             //if(optopt=='H' || optopt=='P' || optopt=='D' || optopt=='M')
             exit(1);
@@ -108,10 +121,10 @@ int main(int argc, char** argv) {
     atexit(&exit_hook);
 
     if(opt_host){
-        server = new RLServer(data_dir, hostaddr, port, db_num);
+        server = new RLServer(data_dir, hostaddr, port, db_num, write_buffer_size);
         server->start();
     }else{
-        server = new RLServer(data_dir, "", port, db_num);
+        server = new RLServer(data_dir, "", port, db_num, write_buffer_size);
         server->start();
     }
 
diff --git a/src/rl_connection.cpp b/src/rl_connection.cpp
index 5e49310..184f2a8 100644
--- a/src/rl_connection.cpp
+++ b/src/rl_connection.cpp
@@ -28,7 +28,7 @@
             read_buffer[buffered_data]=0;                                        \
             return 0;}}while(0)
 
-#define START_WRITER() do{if(!writer_started){writer_started=true;ev_io_start(server->loop, &write_watcher);}}while(0)
+#define START_WRITER() do{if(!writer_started){writer_started=true;ev_io_start(loop, &write_watcher);}}while(0)
 
 RLConnection::RLConnection(RLServer *s, int fd):
     db_index(0), fd(fd), server(s), buffered_data(0),writer_started(false)
@@ -50,26 +50,45 @@ RLConnection::RLConnection(RLServer *s, int fd):
     server->clients_num++;
     //memcpy(sockaddr, &addr, addr_len);
     //ip = inet_ntoa(sockaddr.sin_addr);
+
+    loop = ev_loop_new(0);
 }
 
 RLConnection::~RLConnection()
 {
     if(open){
-        ev_io_stop(server->loop, &read_watcher);
+        ev_io_stop(loop, &read_watcher);
         if(writer_started){
-            ev_io_stop(server->loop, &write_watcher);
+            ev_io_stop(loop, &write_watcher);
         }
         close(fd);
     }
+    if(current_request){
+        delete current_request;
+        current_request=NULL;
+    }
+    if(transaction){
+        delete transaction;
+        transaction=NULL;
+    }
+    ev_loop_destroy(loop);
     server->clients_num--;
 }
 
 void RLConnection::start()
 {
-    ev_io_set(&write_watcher, fd, EV_WRITE);
+    connection_thread = std::thread(&RLConnection::event_loop, this);
+    connection_thread.detach();
+}
+
+void RLConnection::event_loop(RLConnection* conn)
+{
+    ev_io_set(&conn->write_watcher, conn->fd, EV_WRITE);
+
+    ev_io_set(&conn->read_watcher, conn->fd, EV_READ);
+    ev_io_start(conn->loop, &conn->read_watcher);
 
-    ev_io_set(&read_watcher, fd, EV_READ);
-    ev_io_start(server->loop, &read_watcher);
+    ev_run(conn->loop, 0);
 }
 
 
@@ -222,14 +241,14 @@ int RLConnection::do_write(){
         }else{
             perror("Write Error Msg");
             writer_started=false;
-            ev_io_stop(server->loop, &write_watcher);
+            ev_io_stop(loop, &write_watcher);
             return -1;
         }
     }
     write_buffer.erase(0,nwritten);
     if(write_buffer.size()<=0){
         writer_started=false;
-        ev_io_stop(server->loop, &write_watcher);
+        ev_io_stop(loop, &write_watcher);
     }
     return 1;
 }
diff --git a/src/rl_connection.h b/src/rl_connection.h
index 4e63d4a..fbc2beb 100644
--- a/src/rl_connection.h
+++ b/src/rl_connection.h
@@ -10,6 +10,7 @@
 
 #include <vector>
 #include <string>
+#include <thread>
 
 #include <ev.h>
 #include <leveldb/c.h>
@@ -41,6 +42,9 @@ public:
     ev_timer timeout_watcher;
     ev_timer goodbye_watcher;
 
+    struct ev_loop* loop;
+    std::thread connection_thread;
+
     /*** methods ***/
 
     RLConnection(RLServer *s, int fd);
@@ -49,6 +53,8 @@ public:
     static void on_readable(struct ev_loop *loop, ev_io *watcher, int revents);
     static void on_writable(struct ev_loop *loop, ev_io *watcher, int revents);
 
+    static void event_loop(RLConnection *connection);
+
     void start();
     size_t get_int();
     int  do_read();
diff --git a/src/rl_kv.cpp b/src/rl_kv.cpp
index 9d4c5c4..e50d15c 100644
--- a/src/rl_kv.cpp
+++ b/src/rl_kv.cpp
@@ -164,6 +164,32 @@ void RLRequest::rl_set(){
     }
 }
 
+void RLRequest::rl_getset(){
+    if(args.size()!=2){
+        connection->write_error("ERR wrong number of arguments for 'set' command");
+        return;
+    }
+
+    std::string out;
+    leveldb::Status status_get;
+    status_get = connection->server->db[connection->db_index]->Get(connection->server->read_options,
+             args[0], &out);
+
+    leveldb::Status status_set;
+    status_set = connection->server->db[connection->db_index]->Put(connection->server->write_options,
+             args[0], args[1]);
+
+    if(!status_get.ok() && !status_get.IsNotFound()) {
+        connection->write_error("GETSET ERROR 1");
+    } else if(!status_set.ok()) {
+        connection->write_error("GETSET ERROR 2");
+    } else if(status_get.IsNotFound()) {
+        connection->write_nil();
+    } else if(status_get.ok()) {
+        connection->write_bulk(out);
+    }
+}
+
 void RLRequest::rl_del(){
 
     if(args.size()!=1){
diff --git a/src/rl_request.cpp b/src/rl_request.cpp
index b9007a7..8069af2 100644
--- a/src/rl_request.cpp
+++ b/src/rl_request.cpp
@@ -37,6 +37,7 @@ void RLRequest::init_cmd_map()
     RLRequest::cmd_map["keys"]       = &RLRequest::rl_keys;
     RLRequest::cmd_map["select"]     = &RLRequest::rl_select;
     RLRequest::cmd_map["shutdown"]   = &RLRequest::rl_shutdown;
+    RLRequest::cmd_map["ping"]       = &RLRequest::rl_ping;
     RLRequest::cmd_map["multi"]      = &RLRequest::rl_multi;
     RLRequest::cmd_map["exec"]       = &RLRequest::rl_exec;
     RLRequest::cmd_map["discard"]    = &RLRequest::rl_discard;
@@ -47,6 +48,7 @@ void RLRequest::init_cmd_map()
     RLRequest::cmd_map["get"]        = &RLRequest::rl_get;
     RLRequest::cmd_map["set"]        = &RLRequest::rl_set;
     RLRequest::cmd_map["del"]        = &RLRequest::rl_del;
+    RLRequest::cmd_map["getset"]     = &RLRequest::rl_getset;
     RLRequest::cmd_map["mget"]       = &RLRequest::rl_mget;
     RLRequest::cmd_map["mset"]       = &RLRequest::rl_mset;
 
@@ -171,7 +173,7 @@ void RLRequest::rl_select(){
     connection->write_status("OK");
 }
 
-void RLRequest::rl_multi(){
+void RLRequest::rl_multi(){ // TODO: no longer correct after multithreading
     if(connection->transaction){
         connection->write_error("ERR MULTI calls can not be nested");
         return;
@@ -304,3 +306,13 @@ void RLRequest::rl_shutdown(){
     fprintf(stderr, "SHUTDOWN by client\n");
     exit(0);
 }
+
+void RLRequest::rl_ping(){
+    if (args.size()==0) {
+        connection->write_bulk("PONG");
+    } else if (args.size()==1) {
+        connection->write_bulk(args[0]);
+    } else {
+        connection->write_error("ERR wrong number of arguments for 'ping' command");
+    }
+}
diff --git a/src/rl_request.h b/src/rl_request.h
index fb93138..1b36a5b 100644
--- a/src/rl_request.h
+++ b/src/rl_request.h
@@ -42,6 +42,7 @@ public:
     void rl_keys();
     void rl_info();
     void rl_shutdown();
+    void rl_ping();
 
     void rl_multi();
     void rl_exec();
@@ -51,6 +52,7 @@ public:
     void rl_get();
     void rl_set();
     void rl_del();
+    void rl_getset();
 
     void rl_mget();
     void rl_mset();
diff --git a/src/rl_server.cpp b/src/rl_server.cpp
index 65c97ab..663b036 100644
--- a/src/rl_server.cpp
+++ b/src/rl_server.cpp
@@ -11,12 +11,16 @@
 
 #include <unistd.h>
 #include <sys/socket.h>
+#include <sys/un.h>
 #include <netinet/tcp.h> /* TCP_NODELAY */
 #include <netinet/in.h>  /* inet_ntoa */
 #include <arpa/inet.h>   /* inet_ntoa */
 
 #include <leveldb/filter_policy.h>
 
+#include <chrono>
+#include <iostream>
+
 #include "rl_util.h"
 #include "rl_server.h"
 #include "rl_connection.h"
@@ -33,14 +37,16 @@
 #define EVBACKEND EVBACKEND_KQUEUE
 #endif
 
-RLServer::RLServer(const char *_db_path, const char *_hostaddr, int _port, int dbn):
+RLServer::RLServer(const char *_db_path, const char *_hostaddr, int _port, int dbn, size_t write_buffer_size):
     db_num(dbn), db_path(_db_path), hostaddr(_hostaddr), port(_port),
-    fd(-1), clients_num(0)
+    tcp_fd(-1), unix_fd(-1), clients_num(0)
 {
 
     printf("HOSTADDR:(%s)\n",hostaddr.c_str());
     printf("DB_PATH:(%s)\n",db_path.c_str());
     printf("PORT:(%d)\n",port);
+    printf("WRITE_BUFFER_SIZE:(%ld)\n",write_buffer_size);
+    printf("UNIX Socket: /tmp/redis-leveldb.sock\n");
 
     leveldb::Status status;
 
@@ -48,12 +54,20 @@ RLServer::RLServer(const char *_db_path, const char *_hostaddr, int _port, int d
         options = new leveldb::Options[1];
         options[0].create_if_missing = true;
         options[0].filter_policy = leveldb::NewBloomFilterPolicy(10);
+        options[0].write_buffer_size = write_buffer_size;
 
         db=new leveldb::DB*[1];
-        status = leveldb::DB::Open(options[0], db_path.c_str(), &db[0]);
-        if(!status.ok()) {
+        for (int tries = 0; tries < 1000; tries++) {
+            printf("trying to open db for %ld times\n", tries);
+            status = leveldb::DB::Open(options[0], db_path.c_str(), &db[0]);
+            if(status.ok()) break;
+            usleep(10*1000); // 10ms
+        }
+        if(!status.ok()){
             puts("leveldb open error");
             exit(1);
+        }else{
+            puts("leveldb open success");
         }
     }else{
         options = new leveldb::Options[db_num];
@@ -63,6 +77,7 @@ RLServer::RLServer(const char *_db_path, const char *_hostaddr, int _port, int d
         for(int i=0;i<db_num;i++){
             options[i].create_if_missing = true;
             options[i].filter_policy = leveldb::NewBloomFilterPolicy(16);
+            options[i].write_buffer_size = write_buffer_size;
 
             int count = sprintf(buf, "/db-%03d", i);
             //TODO the db path
@@ -76,8 +91,8 @@ RLServer::RLServer(const char *_db_path, const char *_hostaddr, int _port, int d
     }
 
     loop = ev_loop_new(EVBACKEND);
-    connection_watcher.data = this;
-
+    tcp_connection_watcher.data = this;
+    unix_connection_watcher.data = this;
 }
 
 RLServer::~RLServer(){
@@ -93,30 +108,32 @@ RLServer::~RLServer(){
     delete[] options;
     delete[] db;
     if(loop) ev_loop_destroy(loop);
-    close(fd);
+    close(tcp_fd);
+    close(unix_fd);
 }
 
 
 void RLServer::start()
 {
+    // -------------------------------- tcp socket ----------------------------
     struct linger ling = {0, 0};
     struct sockaddr_in addr;
     int flags = 1;
 
-    if((fd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
+    if((tcp_fd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
         perror("socket()");
         exit(1);
     }
 
     flags = 1;
-    setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (void *)&flags, sizeof(flags));
-    setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (void *)&flags, sizeof(flags));
-    setsockopt(fd, SOL_SOCKET, SO_LINGER, (void *)&ling, sizeof(ling));
+    setsockopt(tcp_fd, SOL_SOCKET, SO_REUSEADDR, (void *)&flags, sizeof(flags));
+    setsockopt(tcp_fd, SOL_SOCKET, SO_KEEPALIVE, (void *)&flags, sizeof(flags));
+    setsockopt(tcp_fd, SOL_SOCKET, SO_LINGER, (void *)&ling, sizeof(ling));
 
     /* XXX: Sending single byte chunks in a response body? Perhaps there is a
      * need to enable the Nagel algorithm dynamically. For now disabling.
      */
-    setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (void *)&flags, sizeof(flags));
+    setsockopt(tcp_fd, IPPROTO_TCP, TCP_NODELAY, (void *)&flags, sizeof(flags));
 
     /* the memset call clears nonstandard fields in some impementations that
      * otherwise mess things up.
@@ -135,51 +152,95 @@ void RLServer::start()
         addr.sin_addr.s_addr = htonl(INADDR_ANY);
     }
 
-    if(bind(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+    if(bind(tcp_fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
         perror("bind()");
-        if(fd > 0) close(fd);
+        if(tcp_fd > 0) close(tcp_fd);
         exit(1);
     }
 
-    if(listen(fd, MAX_CONNECTIONS) < 0) {
+    if(listen(tcp_fd, MAX_CONNECTIONS) < 0) {
         perror("listen()");
         exit(1);
     }
 
-    set_nonblock(fd);
+    set_nonblock(tcp_fd);
 
-    ev_init(&connection_watcher, RLServer::on_connection);
+    ev_init(&tcp_connection_watcher, RLServer::on_tcp_connection);
 
-    ev_io_set(&connection_watcher, fd, EV_READ);
-    ev_io_start(loop, &connection_watcher);
+    ev_io_set(&tcp_connection_watcher, tcp_fd, EV_READ);
+    ev_io_start(loop, &tcp_connection_watcher);
 
-    printf("Server running successfully\n");
-    ev_run(loop, 0);
-}
+    // -------------------------------- unix socket ---------------------------
 
-void RLServer::on_connection(struct ev_loop *loop, ev_io *watcher, int revents)
-{
-    RLServer *s = static_cast<RLServer*>(watcher->data);
-    if(EV_ERROR & revents) {
-        puts("on_connection() got error event, closing server.");
-        return;
+    struct sockaddr_un unix_addr;
+    if((unix_fd = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        perror("socket()");
+        exit(1);
     }
 
-    struct sockaddr_in addr; // connector's address information
-    socklen_t addr_len = sizeof(addr);
-    int fd = accept(s->fd, (struct sockaddr*)&addr, &addr_len);
+    memset(&unix_addr, 0, sizeof(unix_addr));
+    unix_addr.sun_family = AF_UNIX;
+    strncpy(unix_addr.sun_path, "/tmp/redis-leveldb.sock", sizeof(unix_addr.sun_path) - 1);
 
-    if(fd < 0) {
-        perror("accept()");
-        return;
-    }
+    unlink("/tmp/redis-leveldb.sock"); // Remove the socket file if it already exists
 
-    RLConnection *connection = new RLConnection(s, fd);
+    if(bind(unix_fd, (struct sockaddr *)&unix_addr, sizeof(unix_addr)) < 0) {
+        perror("bind()");
+        if(unix_fd > 0) close(unix_fd);
+        exit(1);
+    }
 
-    if(connection == NULL) {
-        close(fd);
-        return;
+    if(listen(unix_fd, MAX_CONNECTIONS) < 0) {
+        perror("listen()");
+        exit(1);
     }
-    connection->start();
 
+    set_nonblock(unix_fd);
+
+    ev_init(&unix_connection_watcher, RLServer::on_unix_connection);
+    ev_io_set(&unix_connection_watcher, unix_fd, EV_READ);
+    ev_io_start(loop, &unix_connection_watcher);
+
+    // -------------------------------- ev loop -------------------------------
+
+    printf("Server running successfully\n");
+    const auto now = std::chrono::system_clock::now();
+    std::cout << "boot time " << std::chrono::duration_cast<std::chrono::nanoseconds>(
+                 now.time_since_epoch()).count() << std::endl;
+
+    ev_run(loop, 0);
 }
+
+#define on_connection_def(func_name, fd_name)                              \
+  void RLServer::func_name(struct ev_loop *loop, ev_io *watcher,           \
+                           int revents) {                                  \
+    RLServer *s = static_cast<RLServer *>(watcher->data);                  \
+    if (EV_ERROR & revents) {                                              \
+      puts("on_connection() got error event, closing server.");            \
+      return;                                                              \
+    }                                                                      \
+                                                                           \
+    struct sockaddr_in addr; /* connector's address information */         \
+    socklen_t addr_len = sizeof(addr);                                     \
+    int conn_fd = accept(s->fd_name, (struct sockaddr *)&addr, &addr_len); \
+                                                                           \
+    if (conn_fd < 0) {                                                     \
+      perror("accept()");                                                  \
+      return;                                                              \
+    }                                                                      \
+                                                                           \
+    int flags = 1;                                                         \
+    setsockopt(s->fd_name, IPPROTO_TCP, TCP_NODELAY, (void *)&flags,       \
+               sizeof(flags));                                             \
+                                                                           \
+    RLConnection *connection = new RLConnection(s, conn_fd);               \
+                                                                           \
+    if (connection == NULL) {                                              \
+      close(conn_fd);                                                      \
+      return;                                                              \
+    }                                                                      \
+    connection->start();                                                   \
+  }
+
+on_connection_def(on_tcp_connection, tcp_fd);
+on_connection_def(on_unix_connection, unix_fd);
\ No newline at end of file
diff --git a/src/rl_server.h b/src/rl_server.h
index 3356d66..b4af24a 100644
--- a/src/rl_server.h
+++ b/src/rl_server.h
@@ -10,6 +10,7 @@
 
 #include <vector>
 #include <string>
+#include <atomic>
 
 #include <ev.h>
 #include <leveldb/db.h>
@@ -22,19 +23,21 @@ public:
     std::string db_path;
     std::string hostaddr;
     int port;
-    int fd;
-    int clients_num;
+    int tcp_fd, unix_fd;
+    std::atomic<int> clients_num;
     struct ev_loop* loop;
-    ev_io connection_watcher;
+    ev_io tcp_connection_watcher, unix_connection_watcher;
     leveldb::Options *options;
     leveldb::ReadOptions read_options;
     leveldb::WriteOptions write_options;
     leveldb::DB **db;
 
-    RLServer(const char *db_path, const char *hostaddr, int port, int dbn=0);
+    RLServer(const char *db_path, const char *hostaddr, int port, int dbn=0, size_t write_buffer_size=0);
     ~RLServer();
     void start();
-    static void on_connection(struct ev_loop *loop, ev_io *watcher, int revents);
+
+    static void on_tcp_connection(struct ev_loop *loop, ev_io *watcher, int revents);
+    static void on_unix_connection(struct ev_loop *loop, ev_io *watcher, int revents);
 };
 
 
diff --git a/src/rl_util.cpp b/src/rl_util.cpp
index 16bb92a..71794e1 100644
--- a/src/rl_util.cpp
+++ b/src/rl_util.cpp
@@ -18,6 +18,9 @@
 #include <fcntl.h>
 #include <sys/stat.h>
 
+#include <chrono>
+#include <iostream>
+
 #include "rl_util.h"
 #include "rl_server.h"
 
@@ -83,6 +86,9 @@ void sig_term(int signo) {
     if(signo==SIGINT ||
         signo == SIGTERM ||
         signo == SIGQUIT) {
+        const auto now = std::chrono::system_clock::now();
+        std::cout << "crash time " << std::chrono::duration_cast<std::chrono::nanoseconds>(
+                     now.time_since_epoch()).count() << std::endl;
         exit_hook();
         exit(0);
     }
diff --git a/t/Tester.pm b/t/Tester.pm
index 341099f..8295608 100644
--- a/t/Tester.pm
+++ b/t/Tester.pm
@@ -45,7 +45,8 @@ sub start {
 
 sub stop {
     my $self = shift;
-    my $stop_cmd = "kill -INT " . $self->{'pid'};
+    # my $stop_cmd = "kill -INT " . $self->{'pid'};
+    my $stop_cmd = "pgrep -x \"redis-leveldb\" | xargs kill -9";
     qx($stop_cmd) and die "# server stop error!";
     qx(rm -fr $self->{option}{data_dir});
 }
diff --git a/vendor/leveldb b/vendor/leveldb
index a7bff69..99b3c03 160000
--- a/vendor/leveldb
+++ b/vendor/leveldb
@@ -1 +1 @@
-Subproject commit a7bff697baa062c8f6b8fb760eacf658712b611a
+Subproject commit 99b3c03b3284f5886f9ef9a4ef703d57373e61be
